### PR TITLE
MODINV-331 Upgrade to Vertx v3.9.4 (CVE-2019-17640)

### PR DESCRIPTION
**Pull request by OleksiiKuzminov**
- Incremented Vert.X version to 3.9.4 to support vertx-kafka-client
- Fixed depreciations
- Fixed smells from static code analyzer

### CODE REVIEW

**Review by KaterynaSenchenko — APPROVED** [review]

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+## 6.2.0-SNAPSHOT 2020-11-23
```
Comment (julianladisch) [inline NEWS.md:1]:
```suggestion
## 16.2.0-SNAPSHOT 2020-11-23
```

**Review by VRohach — APPROVED** [review]

**Review by bohdan-suprun — APPROVED** [review]
Thanks for the code cleanup, @OleksiiKuzminov!

**Review by RuslanLavrov — APPROVED** [review]

File: pom.xml [hunk 2]
```diff
@@ -173,7 +173,7 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-    <vertx.version>3.5.4</vertx.version>
+    <vertx.version>3.9.4</vertx.version>
```
Comment (julianladisch) [inline pom.xml:176]: This doesn't upgrade all vertx modules:
```
mvn dependency:tree -Dincludes=io.vertx
[INFO] org.folio:mod-inventory:jar:16.2.0-SNAPSHOT
[INFO] +- io.vertx:vertx-core:jar:3.9.4:compile
[INFO] +- io.vertx:vertx-web:jar:3.9.4:compile
[INFO] |  +- io.vertx:vertx-web-common:jar:3.9.4:compile
[INFO] |  +- io.vertx:vertx-auth-common:jar:3.9.4:compile
[INFO] |  \- io.vertx:vertx-bridge-common:jar:3.9.4:compile
[INFO] +- org.folio:mod-source-record-storage-client:jar:4.1.1:compile
[INFO] |  \- org.folio:domain-models-runtime:jar:31.0.2:compile
[INFO] |     +- io.vertx:vertx-dropwizard-metrics:jar:3.9.2:compile
[INFO] |     \- io.vertx:vertx-pg-client:jar:3.9.2:compile
[INFO] |        \- io.vertx:vertx-sql-client:jar:3.9.2:compile
[INFO] +- org.folio:mod-pubsub-client:jar:1.3.3:compile
[INFO] |  \- org.folio:folio-di-support:jar:1.1.0:compile
[INFO] |     \- io.vertx:vertx-service-proxy:jar:3.9.0:compile
[INFO] +- io.vertx:vertx-unit:jar:3.9.4:test
[INFO] \- io.vertx:vertx-codegen:jar:3.9.4:test
```
vertx-dropwizard-metrics and vertx-service-proxy have still an old version, and more vertx module may get affected whenever a dependency is added or changed in mod-inventory pom.xml.
Please consider using [Vert.x Bill of Materials (Maven BOM)](https://github.com/vert-x3/vertx-stack#bills-of-materials-maven-bom) (content: [vertx-dependencies pom.xml](https://github.com/vert-x3/vertx-dependencies/blob/3.9.4/pom.xml#L49-L54)) for io.vertx:* as suggested in the [RMB upgrading notes](https://github.com/folio-org/raml-module-builder/blob/master/doc/upgrading.md#version-300).

File: src/main/java/org/folio/inventory/dataimport/handlers/actions/AbstractInstanceEventHandler.java [hunk 3]
```diff
@@ -48,7 +49,7 @@
   protected HttpClient client;
 
   protected Future<Void> createPrecedingSucceedingTitles(Instance instance, CollectionResourceRepository precedingSucceedingTitlesRepository) {
-    Future<Void> future = Future.future();
+    Promise<Void> future = Promise.promise();
```
Comment (julianladisch) [inline src/main/java/org/folio/inventory/dataimport/handlers/actions/AbstractInstanceEventHandler.java:52]: A promise should not be called future. Can you rename this variable and all similar changes in this pull request?
Learn how Future was split into Future and Promise: https://github.com/folio-org/raml-module-builder/blob/master/doc/futurisation.md

File: src/main/java/org/folio/inventory/dataimport/handlers/actions/CreateInstanceEventHandler.java [hunk 4]
```diff
@@ -75,7 +76,7 @@ public CreateInstanceEventHandler(Storage storage, HttpClient client) {
         Instance mappedInstance = InstanceUtil.jsonToInstance(instanceAsJson);
         addInstance(mappedInstance, instanceCollection)
           .compose(createdInstance -> createPrecedingSucceedingTitles(mappedInstance, precedingSucceedingTitlesRepository).map(createdInstance))
-          .setHandler(ar -> {
+          .onComplete(ar -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/inventory/dataimport/handlers/actions/CreateInstanceEventHandler.java:79]: The pattern
```
.onComplete(ar -> {
  if (ar.succeded()) {
    // do something with ar.result()
  } else {
    // do something with ar.cause()
  }
})
```
could be replaced by the more concise form
```
.onSuccess(result -> {
  // do something with result
})
.onFailure(cause -> {
  // do something with cause
})
```

File: src/main/java/org/folio/inventory/dataimport/handlers/actions/InstanceUpdateDelegate.java [hunk 5]
```diff
@@ -47,12 +47,12 @@ public InstanceUpdateDelegate(Storage storage) {
...
         .compose(existingInstance -> updateInstance(existingInstance, mappedInstance))
         .compose(updatedInstance -> updateInstanceInStorage(updatedInstance, instanceCollection))
-        .setHandler(future::handle);
+        .onComplete(future);
     } catch (Exception e) {
       LOGGER.error("Error updating inventory instance", e);
       future.fail(e);
     }
-    return future;
+    return future.future();
```
Comment (julianladisch) [inline src/main/java/org/folio/inventory/dataimport/handlers/actions/InstanceUpdateDelegate.java:55]: We don't need a promise, we can directly return a Future:
```
    try {
      ...
      return getInstanceById(instanceId, instanceCollection)
        .compose(existingInstance -> updateInstance(existingInstance, mappedInstance))
        .compose(updatedInstance -> updateInstanceInStorage(updatedInstance, instanceCollection));
    } catch (Exception e) {
      LOGGER.error("Error updating inventory instance", e);
      return Future.failedFuture(e);
    }
```

File: src/main/java/org/folio/inventory/dataimport/handlers/actions/InstanceUpdateDelegate.java [hunk 6]
```diff
@@ -84,17 +84,17 @@ private JsonObject retrieveParsedContent(ParsedRecord parsedRecord) {
       LOGGER.error("Error updating instance", e);
       future.fail(e);
     }
-    return future;
+    return future.future();
```
Comment (julianladisch) [inline src/main/java/org/folio/inventory/dataimport/handlers/actions/InstanceUpdateDelegate.java:87]: We don't need a promise, we can directly return a Future:
`return Future.suceededFuture(mergedInstance);`
`return Future.failedFuture(e);`

File: src/main/java/org/folio/inventory/resources/InstancesBatch.java [hunk 7]
```diff
@@ -226,11 +227,11 @@ private void setInstancesIdIfNecessary(List<JsonObject> instances) {
...
         }
       }
-      CompositeFuture.join(updateRelationshipsFutures).setHandler(resultFuture);
+      CompositeFuture.join(updateRelationshipsFutures).onComplete(resultFuture);
     } catch (IllegalStateException e) {
       log.error("Can not update instances relationships cause: " + e);
       resultFuture.fail(e);
     }
-    return resultFuture;
+    return resultFuture.future();
```
Comment (julianladisch) [inline src/main/java/org/folio/inventory/resources/InstancesBatch.java:235]: There is no need for a promise, we can directly return the Future:
`return CompositeFuture.join(updateRelationshipsFutures);`
`return Future.failedFuture(e);`

**Comment by julianladisch** [comment]
I also appreciate the code cleanup; it helps code review if the changes needed for upgrading dependencies are in one pull request and unrelated code cleanup is in a separate pull request.

File: NEWS.md [hunk 8]
```diff
@@ -1,3 +1,6 @@
...
+* [MODINV-331](https://issues.folio.org/browse/MODINV-331) Upgrade to Vertx v3.9.4 (CVE-2019-17640)
+
```
Comment (marcjohnson-kint) [inline NEWS.md:3]: Why is this news being added in this pull request?

File: NEWS.md (hunk 1 above)
Comment (marcjohnson-kint) [inline NEWS.md:1]: Has this version been released already? If not, why does it have a release date in the past?

**Pull request merged**